File: src/slowql/rules/performance/memory.py
"""
Performance Memory rules.
"""

from __future__ import annotations

from typing import Any

from sqlglot import exp

from slowql.core.models import Category, Dimension, Fix, Issue, Query, Severity
from slowql.rules.base import ASTRule, PatternRule, Rule

__all__ = [
    "GroupByHighCardinalityRule",
    "ImplicitConversionInJoinRule",
    "LargeInClauseRule",
    "OrderByWithoutLimitInSubqueryRule",
    "SelectIntoTempWithoutIndexRule",
    "UnboundedTempTableRule",
]


class LargeInClauseRule(ASTRule):
    """Detects IN clauses with excessive values (>50)."""

    id = "PERF-MEM-001"
    name = "Large IN Clause"
    description = "Detects IN clauses with excessive values (>50) that can cause memory pressure and poor plans."
    severity = Severity.MEDIUM
    dimension = Dimension.PERFORMANCE
    category = Category.PERF_MEMORY

    def check_ast(self, query: Query, ast: Any) -> list[Issue]:
        issues = []

        for node in ast.walk():
            if isinstance(node, exp.In):
                values = getattr(node, "expressions", [])
                if not values and getattr(node, "query", None):
                    continue  # Subquery, not literal list

                if len(values) > 50:
                    issues.append(
                        self.create_issue(
                            query=query,
                            message=f"IN clause with {len(values)} values - consider using temp table or table-valued parameter",
                            snippet=str(node)[:100],
                            impact=(
                                "Large IN clauses (100+ values) consume memory for query compilation, bloat the plan cache "
                                "with unique plans, and may force suboptimal execution strategies."
                            ),
                            fix=Fix(
                                description="Load values into a temp table or table-valued parameter (TVP), then JOIN.",
                                replacement="",
                                is_safe=False,
                            ),
                        )
                    )

        return issues


class UnboundedTempTableRule(PatternRule):
    """Detects SELECT INTO temp table without WHERE clause or row limit."""

    id = "PERF-MEM-002"
    name = "Unbounded Temp Table Creation"
    description = "Detects SELECT INTO temp table without WHERE clause or row limit."
    severity = Severity.MEDIUM
    dimension = Dimension.PERFORMANCE
    category = Category.PERF_MEMORY
    dialects = ("tsql",)

    pattern = r"\bSELECT\b(?!.*\b(WHERE|TOP|LIMIT)\b)[^;]*\bINTO\s+[#@\w]+"
    message_template = "Unbounded SELECT INTO temp table detected: {match}"

    impact = (
        "Unbounded SELECT INTO can fill tempdb, crash the instance, or exhaust memory. "
        "A single runaway query can impact all database users."
    )
    fix_guidance = "Always add WHERE clause or TOP/LIMIT to bound result size. Pre-create temp table with explicit schema for better memory estimation."


class OrderByWithoutLimitInSubqueryRule(ASTRule):
    """Detects ORDER BY in subqueries without LIMIT."""

    id = "PERF-MEM-003"
    name = "ORDER BY Without LIMIT in Subquery"
    description = "Detects ORDER BY in subqueries/CTEs without TOP/LIMIT, which is meaningless and wastes resources."
    severity = Severity.LOW
    dimension = Dimension.PERFORMANCE
    category = Category.PERF_MEMORY

    def check_ast(self, query: Query, ast: Any) -> list[Issue]:
        issues = []

        for node in ast.walk():
            if isinstance(node, exp.Subquery):
                inner = node.this
                if isinstance(inner, exp.Select):
                    has_order = inner.args.get("order") is not None
                    has_limit = inner.args.get("limit") is not None

                    if has_order and not has_limit:
                        issues.append(
                            self.create_issue(
                                query=query,
                                message="ORDER BY in subquery without LIMIT is meaningless and wastes resources",
                                snippet=str(inner)[:100],
                                impact=(
                                    "Sorting requires memory allocation and CPU. ORDER BY without LIMIT in subqueries "
                                    "does nothing (SQL standard ignores it) but still consumes resources."
                                ),
                                fix=Fix(
                                    description="Remove ORDER BY from subqueries unless paired with TOP/LIMIT. Apply ordering in the final outer query only.",
                                    replacement="",
                                    is_safe=False,
                                ),
                            )
                        )

        return issues


class GroupByHighCardinalityRule(ASTRule):
    """Detects GROUP BY on columns likely to have high cardinality (timestamps, IDs, UUIDs)."""

    id = "PERF-MEM-004"
    name = "GROUP BY on High-Cardinality Expression"
    description = (
        "Detects GROUP BY on columns likely to have high cardinality (timestamps, IDs, UUIDs)."
    )
    severity = Severity.LOW
    dimension = Dimension.PERFORMANCE
    category = Category.PERF_MEMORY

    def check_ast(self, query: Query, ast: Any) -> list[Issue]:
        issues = []

        high_cardinality_patterns = {
            "timestamp",
            "datetime",
            "created_at",
            "updated_at",
            "modified_at",
            "uuid",
            "guid",
            "id",
            "transaction_id",
            "session_id",
            "request_id",
            "email",
            "phone",
            "ip_address",
            "user_agent",
        }

        for node in ast.walk():
            if isinstance(node, exp.Select):
                group = node.args.get("group")
                if group:
                    expressions = getattr(group, "expressions", [group])
                    for expr in expressions:
                        if isinstance(expr, exp.Column):
                            col_name = getattr(expr, "name", "").lower()
                            if any(hc in col_name for hc in high_cardinality_patterns):
                                issues.append(
                                    self.create_issue(
                                        query=query,
                                        message=f"GROUP BY on high-cardinality column '{expr.name}' - may create excessive groups",
                                        snippet=str(node)[:100],
                                        impact=(
                                            "Grouping by high-cardinality columns (timestamps, UUIDs) creates millions of groups, "
                                            "consuming massive memory and producing unusable results."
                                        ),
                                        fix=Fix(
                                            description="Truncate timestamps to meaningful intervals (e.g., DATE_TRUNC). Group by categorical columns.",
                                            replacement="",
                                            is_safe=False,
                                        ),
                                    )
                                )

        return issues


class SelectIntoTempWithoutIndexRule(Rule):
    """Detects SELECT INTO # temp table without subsequent index creation."""

    id = "PERF-TSQL-002"
    name = "SELECT INTO Temp Table Without Index"
    description = "SELECT INTO creates a temp table without indexes, causing table scans."
    severity = Severity.MEDIUM
    dimension = Dimension.PERFORMANCE
    category = Category.PERF_MEMORY
    dialects = ("tsql",)
    impact = "Temp tables without indexes cause table scans on every join or filter."
    fix_guidance = "Add CREATE INDEX after SELECT INTO, or pre-create the table with indexes."

    def check(self, query: Query) -> list[Issue]:
        if not self._dialect_matches(query):
            return []
        if not self._has_pattern(query.raw, r"\bSELECT\b.*\bINTO\s+#"):
            return []
        raw_upper = query.raw.upper()
        if "CREATE INDEX" in raw_upper or "CREATE UNIQUE INDEX" in raw_upper:
            return []
        return [self.create_issue(query=query, message="SELECT INTO temp table without index — subsequent queries will table scan.", snippet=query.raw[:80])]


class ImplicitConversionInJoinRule(PatternRule):
    """Detects potential implicit conversion in JOIN predicates in T-SQL."""

    id = "PERF-TSQL-003"
    name = "Implicit Conversion in JOIN Predicate"
    description = "Explicit CONVERT/CAST in JOIN suggests type mismatch preventing index usage."
    severity = Severity.HIGH
    dimension = Dimension.PERFORMANCE
    category = Category.PERF_JOIN
    dialects = ("tsql",)
    pattern = r"\bJOIN\b.*\bCONVERT\s*\(|\bCAST\s*\(.*\bJOIN\b"
    message_template = "Explicit CONVERT/CAST in JOIN predicate — check for implicit conversion: {match}"
    impact = "Implicit conversion prevents index seeks, forcing table scans."
    fix_guidance = "Ensure JOIN columns have matching data types at the schema level."

File: slowql-rs/src/rules/performance/memory.rs
use crate::models::issue::Category;
use crate::models::{Dimension, Issue, Query, Severity};
use crate::rules::base::{DialectSet, Rule};
use once_cell::sync::Lazy;
use regex::Regex;

struct LargeInClauseRule;
static PAT_IN_CLAUSE: Lazy<Regex> = Lazy::new(|| Regex::new(r"(?i)\bIN\s*\(").unwrap());
impl Rule for LargeInClauseRule {
    fn id(&self) -> &'static str { "PERF-MEM-001" }
    fn name(&self) -> &'static str { "Large IN Clause" }
    fn severity(&self) -> Severity { Severity::Medium }
    fn dimension(&self) -> Dimension { Dimension::Performance }
    fn category(&self) -> Option<Category> { Some(Category::PerfMemory) }
    fn impact(&self) -> &'static str { "Large IN clauses (100+ values) consume memory and bloat the plan cache." }
    fn check(&self, query: &Query) -> Vec<Issue> {
        // Count commas inside IN(...) as proxy for value count
        if let Some(m) = PAT_IN_CLAUSE.find(&query.raw) {
            let after = &query.raw[m.end()..];
            if let Some(close) = after.find(')') {
                let inner = &after[..close];
                // Skip if it contains SELECT (subquery, not literal list)
                if inner.to_uppercase().contains("SELECT") { return Vec::new(); }
                let comma_count = inner.matches(',').count();
                if comma_count > 50 {
                    let msg = format!("IN clause with {} values - consider using temp table.", comma_count + 1);
                    return vec![self.build_issue(query, &msg, &query.raw[..query.raw.len().min(100)])];
                }
            }
        }
        Vec::new()
    }
}

struct UnboundedTempTableRule;
static PAT_TEMP: Lazy<Regex> = Lazy::new(|| Regex::new(r"(?i)\bSELECT\b[^;]*\bINTO\s+[#@]\w+").unwrap());
impl Rule for UnboundedTempTableRule {
    fn id(&self) -> &'static str { "PERF-MEM-002" }
    fn name(&self) -> &'static str { "Unbounded Temp Table Creation" }
    fn severity(&self) -> Severity { Severity::Medium }
    fn dimension(&self) -> Dimension { Dimension::Performance }
    fn category(&self) -> Option<Category> { Some(Category::PerfMemory) }
    fn dialects(&self) -> DialectSet { DialectSet::new(&["tsql"]) }
    fn impact(&self) -> &'static str { "Unbounded SELECT INTO can fill tempdb, crash the instance, or exhaust memory." }
    fn check(&self, query: &Query) -> Vec<Issue> {
        if !self.dialect_matches(query) { return Vec::new(); }
        if let Some(m) = PAT_TEMP.find(&query.raw) {
            let upper = query.raw_upper();
            if upper.contains("WHERE") || upper.contains("TOP ") || upper.contains("LIMIT") {
                return Vec::new();
            }
            let msg = format!("Unbounded SELECT INTO temp table: {}", m.as_str());
            return vec![self.build_issue(query, &msg, m.as_str())];
        }
        Vec::new()
    }
}

struct GroupByHighCardinalityRule;
static HIGH_CARD_COLS: &[&str] = &[
    "timestamp", "datetime", "created_at", "updated_at", "modified_at",
    "uuid", "guid", "transaction_id", "session_id", "request_id",
    "email", "phone", "ip_address", "user_agent",
];
impl Rule for GroupByHighCardinalityRule {
    fn id(&self) -> &'static str { "PERF-MEM-004" }
    fn name(&self) -> &'static str { "GROUP BY on High-Cardinality Expression" }
    fn severity(&self) -> Severity { Severity::Low }
    fn dimension(&self) -> Dimension { Dimension::Performance }
    fn category(&self) -> Option<Category> { Some(Category::PerfMemory) }
    fn impact(&self) -> &'static str { "Grouping by high-cardinality columns creates millions of groups, consuming massive memory." }
    fn check(&self, query: &Query) -> Vec<Issue> {
        let upper = query.raw_upper();
        if !upper.contains("GROUP BY") { return Vec::new(); }
        let raw_lower = query.raw.to_lowercase();
        for col in HIGH_CARD_COLS {
            if raw_lower.contains(&format!("group by {}", col)) || raw_lower.contains(&format!("group by\n{}", col)) {
                let msg = format!("GROUP BY on high-cardinality column '{}' - may create excessive groups.", col);
                let snip = &query.raw[..query.raw.len().min(100)];
                return vec![self.build_issue(query, &msg, snip)];
            }
        }
        Vec::new()
    }
}

struct SelectIntoTempWithoutIndexRule;
impl Rule for SelectIntoTempWithoutIndexRule {
    fn id(&self) -> &'static str { "PERF-TSQL-002" }
    fn name(&self) -> &'static str { "SELECT INTO Temp Table Without Index" }
    fn severity(&self) -> Severity { Severity::Medium }
    fn dimension(&self) -> Dimension { Dimension::Performance }
    fn category(&self) -> Option<Category> { Some(Category::PerfMemory) }
    fn dialects(&self) -> DialectSet { DialectSet::new(&["tsql"]) }
    fn impact(&self) -> &'static str { "Temp tables without indexes cause table scans on every join or filter." }
    fn check(&self, query: &Query) -> Vec<Issue> {
        if !self.dialect_matches(query) { return Vec::new(); }
        let upper = query.raw_upper();
        if !upper.contains("INTO #") { return Vec::new(); }
        if upper.contains("CREATE INDEX") || upper.contains("CREATE UNIQUE INDEX") { return Vec::new(); }
        let snip = &query.raw[..query.raw.len().min(80)];
        vec![self.build_issue(query, "SELECT INTO temp table without index - subsequent queries will table scan.", snip)]
    }
}

struct ImplicitConversionInJoinRule;
static PAT_CONV_JOIN: Lazy<Regex> = Lazy::new(|| Regex::new(r"(?i)(\bJOIN\b.*\bCONVERT\s*\()|(\bCAST\s*\(.*\bJOIN\b)").unwrap());
impl Rule for ImplicitConversionInJoinRule {
    fn id(&self) -> &'static str { "PERF-TSQL-003" }
    fn name(&self) -> &'static str { "Implicit Conversion in JOIN Predicate" }
    fn severity(&self) -> Severity { Severity::High }
    fn dimension(&self) -> Dimension { Dimension::Performance }
    fn category(&self) -> Option<Category> { Some(Category::PerfJoin) }
    fn dialects(&self) -> DialectSet { DialectSet::new(&["tsql"]) }
    fn impact(&self) -> &'static str { "Implicit conversion prevents index seeks, forcing table scans." }
    fn check(&self, query: &Query) -> Vec<Issue> {
        if !self.dialect_matches(query) { return Vec::new(); }
        PAT_CONV_JOIN.find(&query.raw).map(|m| {
            let msg = format!("Explicit CONVERT/CAST in JOIN predicate - check for implicit conversion: {}", m.as_str());
            vec![self.build_issue(query, &msg, m.as_str())]
        }).unwrap_or_default()
    }
}

pub fn rules() -> Vec<Box<dyn Rule>> {
    vec![
        Box::new(LargeInClauseRule),
        Box::new(UnboundedTempTableRule),
        Box::new(GroupByHighCardinalityRule),
        Box::new(SelectIntoTempWithoutIndexRule),
        Box::new(ImplicitConversionInJoinRule),
        Box::new(OrderByWithoutLimitInSubqueryRule),
    ]
}

// PERF-MEM-003: ORDER BY Without LIMIT in Subquery
struct OrderByWithoutLimitInSubqueryRule;
static PAT_ORDER_SUB_NO_LIMIT: Lazy<Regex> = Lazy::new(|| Regex::new(r"(?i)\(\s*SELECT\b[^)]*\bORDER\s+BY\b").unwrap());
impl Rule for OrderByWithoutLimitInSubqueryRule {
    fn id(&self) -> &'static str { "PERF-MEM-003" }
    fn name(&self) -> &'static str { "ORDER BY Without LIMIT in Subquery" }
    fn severity(&self) -> Severity { Severity::Low }
    fn dimension(&self) -> Dimension { Dimension::Performance }
    fn category(&self) -> Option<Category> { Some(Category::PerfMemory) }
    fn impact(&self) -> &'static str { "Sorting without LIMIT in subqueries wastes resources." }
    fn check(&self, query: &Query) -> Vec<Issue> {
        if let Some(m) = PAT_ORDER_SUB_NO_LIMIT.find(&query.raw) {
            let matched_upper = m.as_str().to_uppercase();
            if !matched_upper.contains("LIMIT") && !matched_upper.contains("TOP") {
                return vec![self.build_issue(query, "ORDER BY in subquery without LIMIT is meaningless and wastes resources.", m.as_str())];
            }
        }
        Vec::new()
    }
}
